service-proxy.js

```diff
@@ -494,60 +494,80 @@ ServiceDispatchHandler.prototype.computePartialRange =
 function computePartialRange(serviceName, hostPort) {
     var self = this;
 
-    // Obtain and sort the affine worker and relay lists.
-    var relays = Object.keys(self.egressNodes.exitsFor(serviceName));
-    relays.sort();
     var serviceChannel = self.getOrCreateServiceChannel(serviceName);
-    var workers = serviceChannel.peers.keys();
-    workers.sort();
+
+    var range = {
+        relays: null,
+        workers: null,
+        relayIndex: NaN,
+        ratio: NaN,
+        start: NaN,
+        stop: NaN,
+        length: NaN,
+        affineWorkers: []
+    };
+
+    // Obtain and sort the affine worker and relay lists.
+    range.relays = Object.keys(self.egressNodes.exitsFor(serviceName));
+    range.relays.sort();
+    range.workers = serviceChannel.peers.keys();
+    range.workers.sort();
 
     // Find our position within the affine relay set so we can project that
     // position into the affine worker set.
-    var relayIndex = sortedIndexOf(relays, self.channel.hostPort);
+    range.relayIndex = sortedIndexOf(range.relays, self.channel.hostPort);
     // istanbul ignore if
-    if (relayIndex < 0) {
+    if (range.relayIndex < 0) {
         // This should only occur if an advertisement loses the race with a
         // relay ring membership change.
-        return {
-            start: -1,
-            stop: -1,
-            relayIndex: relayIndex,
-            relays: relays,
-            workers: workers,
-            length: -1
-        };
+        return range;
     }
 
     // Compute the range of workers that this relay should be connected to.
-    var ratio = workers.length / relays.length;
-    var start = Math.floor(relayIndex * ratio);
-    var length = Math.ceil(
+    range.ratio = range.workers.length / range.relays.length;
+    range.start = Math.floor(range.relayIndex * range.ratio);
+    range.length = Math.ceil(
         Math.min(
-            workers.length,
+            range.workers.length,
             Math.max(
                 self.minPeersPerRelay,
-                self.minPeersPerWorker * ratio
+                self.minPeersPerWorker * range.ratio
             )
         )
     );
-    var stop = (start + length) % workers.length;
-
-    return {
-        start: start,
-        stop: stop,
-        relayIndex: relayIndex,
-        relays: relays,
-        workers: workers,
-        length: length
-    };
+    range.stop = (range.start + range.length) % range.workers.length;
+
+    if (range.start === range.stop) {
+        // fully connected
+        range.affineWorkers = range.workers; // XXX .slice(0)?
```

Comment:
.slice() would be superfluous. We will treat the array as immutable. Can remove comment.